File: solutions/javascript/hamming/1/hamming.js
//
// This is only a SKELETON file for the 'Hamming' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class UnequalLengthStrands extends Error {
  constructor() {
    super("strands must be of equal length");
  }
}

export const compute = (strand1, strand2) => {
  if (strand1.length != strand2.length) 
    throw new UnequalLengthStrands;
  return [...strand1].reduce((acc, nucleotide, idx) => {
    return acc + (nucleotide != strand2[idx]);
  }, 0);
}

File: solutions/cpp/hamming/1/hamming.cpp
#include "hamming.h"

namespace hamming {
    auto compute(const std::string& strand1, const std::string& strand2) -> uint64_t {
        if (strand1.size() != strand2.size()) {
            throw std::domain_error("strands of different sizes.");
        }

        return std::transform_reduce(strand1.cbegin(), strand1.cend(), strand2.cbegin(), 
            uint64_t{}, std::plus{}, [] (auto nucleotide1, auto nucleotide2) {
                return nucleotide1 != nucleotide2;
            });          
    }

}  // namespace hamming
